### PR TITLE
feat: Implement assistant API streaming 

### LINKED ISSUE #266: Implement assistant API streaming
**Is your feature request related to a problem? Please describe.**
OpenAI just launched streaming support in the assistant API. CopilotKit should support streaming responses via the assistant API.

**Describe the solution you'd like**
Modify https://github.com/CopilotKit/CopilotKit/blob/main/CopilotKit/packages/backend/src/lib/openai-assistant-adapter.ts to use streaming. See: https://platform.openai.com/docs/assistants/overview?context=with-streaming

### PR DESCRIPTION
**Pull request by Bhavyajain21**
/claim #266 
/fixes #266

### CODE REVIEW

**Comment by Bhavyajain21** [comment]
Please review @mme / @ataibarkai

**Comment by ataibarkai** [comment]
@Bhavyajain21 the merge checks are failing-

To fix those, please follow the testing guide in the [contributing quickstart](https://docs.copilotkit.ai/contribute/quickstart-contribute).

Thanks!

File: CopilotKit/packages/backend/src/lib/openai-assistant-adapter.ts [hunk 1]
```diff
@@ -117,18 +117,32 @@ export class OpenAIAssistantAdapter implements CopilotKitServiceAdapter {
...
     // get the thread from forwardedProps or create a new one
     const threadId: string =
       forwardedProps.threadId || (await this.openai.beta.threads.create()).id;
+  
+    // Use createAndStream for user messages
+    if (
+      forwardMessages.length > 0 &&
+      forwardMessages[forwardMessages.length - 1].role === "user"
+    ) {
+      const run = await this.openai.beta.threads.runs.createAndStream(threadId, {
```
Comment (ataibarkai) [inline CopilotKit/packages/backend/src/lib/openai-assistant-adapter.ts:138]: you want to make the `createAndStream` change inside the existing `submitUserMessage` function. 

it encapsulates additional context injections and business logic

File: CopilotKit/packages/backend/src/lib/openai-assistant-adapter.ts [hunk 2]
```diff
@@ -139,19 +153,12 @@ export class OpenAIAssistantAdapter implements CopilotKitServiceAdapter {
     ) {
       run = await this.submitToolOutputs(threadId, forwardedProps.runId, forwardMessages);
     }
-    // submit user message
-    else if (
-      forwardMessages.length > 0 &&
-      forwardMessages[forwardMessages.length - 1].role === "user"
-    ) {
-      run = await this.submitUserMessage(threadId, forwardedProps);
```
Comment (ataibarkai) [inline CopilotKit/packages/backend/src/lib/openai-assistant-adapter.ts:147]: per the other comment: we want to keep this code here.
The changes should be made inside the `submitUserMessage` function definition

File: CopilotKit/packages/backend/src/lib/openai-assistant-adapter.ts [hunk 3]
```diff
@@ -177,6 +184,77 @@ export class OpenAIAssistantAdapter implements CopilotKitServiceAdapter {
       };
     }
   }
+
+  async buildInstructions(forwardMessages: Message[]): Promise<string> {
```
Comment (ataibarkai) [inline CopilotKit/packages/backend/src/lib/openai-assistant-adapter.ts:188]: the existing `submitUserMessage` function already performs the instruction-building and tool-building steps clearly enough, there is no need for defining these as separate functions

File: CopilotKit/packages/backend/src/lib/openai-assistant-adapter.ts [hunk 4]
```diff
@@ -177,6 +184,77 @@ export class OpenAIAssistantAdapter implements CopilotKitServiceAdapter {
...
+      ...(forwardedProps.tools || []), 
+      ...(this.codeInterpreterEnabled ? [{ type: "code_interpreter" }] : []),
+      ...(this.retrievalEnabled ? [{ type: "retrieval" }] : []),
+    ];
+  
+    return tools;
+  }
+
+  async handleStreamingRun(run: OpenAI.Beta.Threads.Runs.RunStream) {
+    return new Promise((resolve) => {
```
Comment (ataibarkai) [inline CopilotKit/packages/backend/src/lib/openai-assistant-adapter.ts:208]: this implementation is on the right direction, but I believe we want to use our existing abstractions here,

see `writeChatCompletionChunk`, `writeChatCompletionEnd` and `AssistantSingleChunkReadableStream` for an example.

@mme can you chime in here please?

**Review by ataibarkai — CHANGES_REQUESTED** [review]
Thanks for your changes-

I made a few notes

**Comment by ataibarkai** [comment]
@Bhavyajain21 just wanted to make sure you saw this

**Comment by Bhavyajain21** [comment]
Hey @ataibarkai, will work on the reviews today!

**Pull request closed without merging**